hphp/runtime/vm/jit/retranslate-all.md

```diff
@@ -0,0 +1,108 @@
+Overview
+--------
+
+The retranslate all feature performs many retranslations simultaneously. These
+threads need space to translate into before statically relocate to the final
+location in the translation cache. If we allocated this space for every thread,
+then we'd run out of low memory. Therefore a single piece of low memory is
+allocated to be used as a "virtual address space". This is where the worker
+threads *think* they're writing the translations, when in fact the assembler
+is writing them to the actual location allocated by the worker thread as
+scratch space.
+
+
+X64 usage of toDestAddress()
+----------------------------
+
+Any time the compiler is reading or writing a location, it must use the
+actual address, which is returned by DataBlock::toDestAddress().
+
+The following is a list of the places that currently do this for the x64
+implementation:
+
+-   ~Label() in the X64 Assembler. It patches up any instructions to
+    that label.
+-   relocateImpl() in relocation-x64.cpp. We copy the src addresses
+    to dest addresses one at a time using DataBlock::bytes(). We use
+    toDestAddress() then and when we decode instructions.
+-   emitSmashableMovq() uses toDestAddress when calculating where to
+    write the embedded literal.
+-   toReal() in vasm-x64.cpp looks up the code block and does
+    toDestAddress(). toReal() is used by:
+    -   retargetJumps()
+    -   patch()
+-   emit mcprep{} in vasm-x64.cpp.
+
+
+ARM usage of toDestAddress()
+----------------------------
+
+ARM similarly must use toDestAddress() a fair amount. The following TODOs
+describe this work for ARM.
+
+
+### toDestAddress() TODOs
+
+*   Update Vixl assembler to use toDestAddress() when necessary.
+    *   Update binding/linking with virtual address support. Label::link_ and
+        Label::target_ are virtual addresses. When binding a Label, the actual
+	instructions linked to it will be updated with target built from the
+	offset between the current link_ and target_.
+*   Update relocateImpl() and all of the helpers.
+    *   Use toDestAddress() when casting to Instruction.
+    *   Fix uses of SetImmPCOffset() and ImmPCOffset(). (details in next section)
+*   Update all uses of __builtin__clear_cache().
+    *   align-arm.cpp
+    *   func-guard-arm.cpp
+    *   relocation-arm.cpp
+    *   smashable-instr-arm.cpp
+    x   tc-relocate.cpp
+    *   unique-stubs-arm.cpp
+    *   vasm-arm.cpp
```

Comment:
It looks like a bunch of these were covered in this diff. Are these still TODOs?